pkg/task/task.go

```diff
@@ -102,39 +108,114 @@ func (pp *Profiler) logProfileSummary(taskName string) {
 
 // Manages Task execution
 type taskRunner struct {
-	task   Task
-	writer filewriter.FileWriter
+	task           Task
+	writer         filewriter.FileWriter
+	withCheckpoint bool
+}
+
+type TaskRunnerOpt func(*taskRunner)
+
+func WithCheckpointFile() TaskRunnerOpt {
+	return func(t *taskRunner) {
+		t.withCheckpoint = true
+		if !features.IsActive(features.CheckpointEnabled()) {
+			logger.V(4).Info("CHECKPOINT_ENABLED env var is not set. Continuing operation without this feature...", features.CheckpointEnabledEnvVar, features.IsActive(features.CheckpointEnabled()))
+			t.withCheckpoint = false
+		}
+	}
 }
 
-// executes Task
 func (pr *taskRunner) RunTask(ctx context.Context, commandContext *CommandContext) error {
+	checkpointFileName := fmt.Sprintf("%s-checkpoint.yaml", commandContext.ClusterSpec.Cluster.Name)
+	var checkpointInfo CheckpointInfo
+	var err error
+
 	commandContext.Profiler = &Profiler{
 		metrics: make(map[string]map[string]time.Duration),
 		starts:  make(map[string]map[string]time.Time),
 	}
 	task := pr.task
 	start := time.Now()
 	defer taskRunnerFinalBlock(start)
+
+	if pr.withCheckpoint {
+		checkpointInfo, err = pr.setupCheckpointInfo(commandContext, checkpointFileName)
+		if err != nil {
+			return err
+		}
+	}
+
 	for task != nil {
+		if pr.withCheckpoint {
+			if completedTask, ok := checkpointInfo.CompletedTasks[task.Name()]; ok {
+				logger.V(4).Info("Restoring task", "task_name", task.Name())
+				nextTask, err := task.Restore(ctx, commandContext, completedTask)
+				if err != nil {
+					return fmt.Errorf("restoring checkpoint info: %v", err)
+				}
+				task = nextTask
+				continue
+			}
+		}
 		logger.V(4).Info("Task start", "task_name", task.Name())
 		commandContext.Profiler.SetStartTask(task.Name())
 		nextTask := task.Run(ctx, commandContext)
 		commandContext.Profiler.MarkDoneTask(task.Name())
 		commandContext.Profiler.logProfileSummary(task.Name())
+		if commandContext.OriginalError == nil && pr.withCheckpoint {
+			checkpointInfo.taskCompleted(task.Name(), task.Checkpoint())
+		}
 		task = nextTask
 	}
+	if commandContext.OriginalError != nil && pr.withCheckpoint {
```

Comment:
same about `pr.withCheckpoint`

Reply:
Keeping it here because we shouldn't attempt to save the checkpoint & write it to a file if that feature isn't enabled

Reply:
Ah I see. Why do you think that? Writing to a file doesn't have any consequences at all. It's only reading from file and restoring tasks what affects the behavior of the program (unless I'm missing something).

Reply:
True, there aren't any actual consequences. I didn't want to write to the file because I didn't want to have an unnecessary file in the temp directory if the operation failed, potentially adding confusion from the user's POV if they didn't enable the checkpoint feature. But once we remove the feature flag, this will be the default behavior so I'm ok with removing the bool check here too